env/edit.go

```diff
@@ -20,18 +22,17 @@ var (
 	errInvalidNameFormat = fmt.Errorf("name can only be composed with alphanumerical characters, hyphens and underscores")
 )
 
-func Add(ctx context.Context, app string, params []string) error {
+func Add(ctx context.Context, app string, params []string, filePath string) error {
 	var variables scalingo.Variables
-	for _, param := range params {
-		if err := isEnvEditValid(param); err != nil {
-			return errgo.Newf("'%s' is invalid: %s", param, err)
-		}
 
-		name, value := parseVariable(param)
-		variables = append(variables, &scalingo.Variable{
-			Name:  name,
-			Value: value,
-		})
+	err := readFromFile(&variables, filePath)
+	if err != nil {
+		return err
+	}
+
+	err = readFromCmdLine(&variables, params)
+	if err != nil {
+		return err
```

Comment:
best practice: I would prefer having both `readFromFile` and `readFromCmdLine` returning a `scalingo.Variables` then merging them after these calls. We are not keen on modifying functions input arguments at Scalingo.

best practice: please wrap the returned error with `github.com/Scalingo/go-utils/errors/v2.Notef`

question: what if `.env` contains a variable which is also defined on the command line? Is it defined twice in `variables`? Shouldn't we make sure it's only once in `variables`?

Reply:
All done!

I fixed the case when a variable has been defined twice. The command line variable will take precedence from the env file.